include/oneapi/dpl/internal/scan_by_segment_impl.h

```diff
@@ -0,0 +1,506 @@
+
+        int32_t __n_groups = 1 + std::ceil(__n / (__wgroup_size * __vals_per_item));
+
+        // intermediate scan values stored by each work item
+        auto __partials =
+            oneapi::dpl::__par_backend_hetero::__internal::__buffer<_ExecutionPolicy, __val_type>(__exec, __n_groups)
+                .get_buffer();
+
+        auto __end_idx =
+            oneapi::dpl::__par_backend_hetero::__internal::__buffer<_ExecutionPolicy, __val_type>(__exec, 1)
+                .get_buffer();
+
+        // the number of segment ends found in each work group
+        auto __seg_ends =
+            oneapi::dpl::__par_backend_hetero::__internal::__buffer<_ExecutionPolicy, bool>(__exec, __n_groups)
+                .get_buffer();
+
+        // 1. Work group scan
+        auto __wg_incl_scan = __exec.queue().submit([&](sycl::handler& __cgh) {
+            auto __partials_acc = __partials.template get_access<sycl::access_mode::write>(__cgh);
+            auto __seg_ends_acc = __seg_ends.template get_access<sycl::access_mode::write>(__cgh);
+
+            oneapi::dpl::__ranges::__require_access(__cgh, __keys, __values, __out_values);
+
+            auto __loc_acc = sycl::accessor<__val_type, 1, sycl::access::mode::read_write, sycl::access::target::local>{
+                2 * __wgroup_size, __cgh};
+
+            __cgh.parallel_for(
+                sycl::nd_range<1>{__n_groups * __wgroup_size, __wgroup_size}, [=](sycl::nd_item<1> __item) {
+                    __val_type __accumulator = {};
+
+                    auto __group = __item.get_group();
+                    ::std::size_t __global_id = __item.get_global_id();
+                    ::std::size_t __local_id = __group.get_local_id();
+
+                    // 1a. Perform a serial scan within the work item over assigned elements. Store partial
+                    // reductions in work item memory, and write the accumulated value and number of counted
+                    // segments into work group memory.
+                    int32_t __start = __global_id * __vals_per_item;
+                    int32_t __end = sycl::minimum{}(__start + __vals_per_item, __n);
+
+                    if (__global_id == 0 && __local_id == 0)
+                        __accumulator = {};
+
+                    // Flag indicating if a new segment exists in the work item's search space
+                    int __max_end = 0;
+
+                    bool __first = true;
+
+                    // Edge case for when the first element is the start of a new segment. We do 
+                    // not want to apply a null aggregate since there is no guarantee it is an identity
+                    if (__start < __n && __local_id != 0 && __keys[__start] != __keys[__start - 1])
+                        __max_end = __local_id;
+
+                    for (int32_t __i = __start; __i < __end; ++__i)
+                    {
+                        if (__first)
```

Comment:
I once read that inside loops it is better to compose a condition in such a way that a more frequent situation is performed in the if branch and not in the else branch. I am not sure, is it applicable there, but may be...

Reply:
This is now out of date, as we have removed this branching by explicitly requiring and using a known_identity for the by_segment rewrites.